add 'testing' feature

masonium opened the pull request:
This adds a 'testing' feature, which compiles and links the tmglib library containing all of the matrix generation routines in TESTING/MATGEN.

I thought it appropriate to make 'testing' a default feature, since openblas-src builds these test routines by default.

IvanUkhov commented:
Not sure how it all compiles on Travis, but my understanding is that, if `BUILD_TESTING` is enabled, all three testing directories should be present. So one won’t be able to exclude any of them in `Cargo.toml`, which will make the crate unpublishable as the archive will be more than 10Mb.

However, the above is a secondary issue. The main one is: What is the motivation?

masonium commented:
My motivation is to have access to the routines in MATGEN for generating random matrices in my [linxal](https://github.com/masonium/linxal) library. 

You're right about the testing directories. I'm guessing that travis just checks out all of `source`, rather than respective the `Cargo.toml` package exclusions, so the code is still present. I modified my code to add the `LAPACKE_WITH_TMG` build option instead of the `BUILD_TESTING` option. This specifically only includes `TESTING/MATGEN` in the build. I confirmed this locally by deleting the `CBLAS/TESTING`, `BLAS/TESTING`, and `TESTING/*.in` files before building, and the build was still successful.

I changed the feature name from `testing` to `tmg` as well.

IvanUkhov commented:
I’ve looked at what gets compiled in [`openblas-src`](https://github.com/cmr/openblas-src). It seems that `tmglib` is always present in that case as we compile the [`netlib`](https://github.com/xianyi/OpenBLAS/blob/85636ff1a015d04d3a8f960bc644b85ee5157135/Makefile#L208) target in [`build.rs`](https://github.com/cmr/openblas-src/blob/master/build.rs#L26), and `netlib` pulls [`tmglib`](https://github.com/xianyi/OpenBLAS/blob/85636ff1a015d04d3a8f960bc644b85ee5157135/Makefile#L211). So, for consistency, it makes sense to always compile `tmglib` here as well. No new feature is needed. Regarding the exclude rules, they should resemble the ones in `openblas-src`; the relevant ones are:

```toml
exclude = [
    "source/lapack-netlib/BLAS/TESTING/*",
    "source/lapack-netlib/CBLAS/testing/*",
    "source/lapack-netlib/TESTING/*.in",
    "source/lapack-netlib/TESTING/EIG/*",
    "source/lapack-netlib/TESTING/LIN/*",
]
```

That is, not only `*.in` should be excluded but also `EIG` and `LIN`. Could you please check that it works with excluding those two in addition? Also, could you please squash all the commits into one? See my comment regarding squashing in [`blas`](https://github.com/stainless-steel/blas/pull/19). Thanks!
